File: integrations/csharp/Oicana/Manifest/InputDefinition.cs
using System.Text.Json.Serialization;

namespace Oicana.Manifest;

/// <summary>
/// An input a template declares.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(JsonInputDefinition), "json")]
[JsonDerivedType(typeof(BlobInputDefinition), "blob")]
public abstract class InputDefinition
{
    /// <summary>
    /// Key the input is supplied and used under.
    /// </summary>
    [JsonPropertyName("key")]
    public required string Key { get; init; }

    /// <summary>
    /// Whether a value of this input is required for compilation.
    /// </summary>
    [JsonPropertyName("required")]
    public required bool Required { get; init; }
}

File: integrations/java/oicana/src/main/java/com/oicana/manifest/InputDefinition.java
package com.oicana.manifest;

/**
 * An input a template declares.
 */
public sealed interface InputDefinition permits JsonInputDefinition, BlobInputDefinition {

    /**
     * Key the input is supplied and used under.
     *
     * @return the input key
     */
    String key();

    /**
     * Whether a value of this input is required for compilation.
     *
     * @return whether a value is required
     */
    boolean required();
}
